thermal-control-design.md

```diff
@@ -198,24 +397,67 @@ We don't have a CLI for fan status getting yet, new CLI for fan status could be
 
 	Commands:
-	  fanstatus  Show fan status information
+	  fan        Show fan status information
 	  mlnx       Mellanox platform specific configuration...
 	  psustatus  Show PSU status information
 	  summary    Show hardware platform information
 	  syseeprom  Show system EEPROM information
 The output of the command is like below:
 
-	admin@sonic# show platform fanstatus
-	FAN    Speed      Direction      Timestamp
-	-----  ---------  ---------  -----------------
-	FAN 1  12919 RPM  Intake     20191112 09:38:16
-	FAN 2  13043 RPM  Exhaust    20191112 09:38:16
+	admin@sonic# show platform fan
+	Drawer     FAN    Speed      Direction  Presence     Status  Timestamp
+	--------   -----  ---------  ---------  -----------  ------  -----------------
+	Drawer 1   FAN 1  85%        intake     Present      OK      20191112 09:38:16
+	           FAN 2  60%        intake     Present      OK      20191112 09:38:16
+    Drawer 2   FAN 3  75%        exhaust    Present      Not OK  20191112 09:38:16
+               FAN 4  65%        exhaust    Present      Not OK  20191112 09:38:16
+
+The output for virtual drawer is like below:
+
+   	admin@sonic# show platform fan
+	Drawer     FAN    Speed      Direction  Presence     Status  Timestamp
+	--------   -----  ---------  ---------  -----------  ------  -----------------
+	N/A        FAN 1  85%        intake     Present      OK      20191112 09:38:16
+	N/A        FAN 2  60%        intake     Present      OK      20191112 09:38:16
+    N/A        FAN 3  75%        exhaust    Present      Not OK  20191112 09:38:16
+    N/A        FAN 4  65%        exhaust    Present      Not OK  20191112 09:38:16 
+
+## 5. Fan drawer implementation
+
+In current implementation, fan led management API is based on fan object. However, in many hardware platform, multiple fans are put in a fan drawer, they share one fan led. In such case, to control fan led in a fan object might cause problem. For example, one fan drawer has fan1 and fan2, they share a fan led, fan1 is broken and set led to red via fan1 object, then fan2 just recovers from a bad state and set led to green via fan2 object, the final fan led state will be green which is incorrect. To avoid confusing user, a new abstract class FanDrawerBase is added to sonic_platform_common and vendors should implement it. The FanDrawerBase class will look like:
+
+```python
+
+class FanDrawerBase(device_base.DeviceBase):
+    def get_num_fans(self):
+        raise NotImplementedError
+
+    def get_all_fans(self):
+        raise NotImplementedError
+
+    def set_status_led(self, color):
+        raise NotImplementedError
+
+    def get_status_led(self, color):
+        raise NotImplementedError
+
+    def is_virtual_drawer(self):
+        raise NotImplementedError
+```
+
+The fan drawer object should follow:
+
+- If there are fans without drawer, we still add a virtual drawer object for each of fan, and consider the drawer only has one fan. We don't show virtual drawer in CLI.
```

Comment:
should make it clear that thermalctld go through drawer first, so have to have a virtual drawer object even no drawer.